module/epochs/base_contract_client.go

```diff
@@ -0,0 +1,103 @@
+package epochs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	sdk "github.com/onflow/flow-go-sdk"
+	sdkcrypto "github.com/onflow/flow-go-sdk/crypto"
+
+	"github.com/onflow/flow-go/module"
+)
+
+// BaseContractClient represents the core fields and methods needed to create
+// a client to a contract on the Flow Network.
+type BaseContractClient struct {
+	log zerolog.Logger // default logger
+
+	ContractAddress string           // contract address
+	AccountKeyIndex uint             // account key index
+	Signer          sdkcrypto.Signer // signer used to sign transactions
+	Account         *sdk.Account     // account belonging to node interacting with the contract
+
+	Client module.SDKClientWrapper // flow access node client
+}
+
+// NewBaseContractClient ...
+func NewBaseContractClient(log zerolog.Logger,
+	flowClient module.SDKClientWrapper,
+	accountAddress string,
+	accountKeyIndex uint,
+	signer sdkcrypto.Signer,
+	contractAddress string) (*BaseContractClient, error) {
+
+	// get account from access node for given address
+	account, err := flowClient.GetAccount(context.Background(), sdk.HexToAddress(accountAddress))
+	if err != nil {
+		return nil, fmt.Errorf("could not get account: %w", err)
+	}
```

Comment:
I like this idea of failing in the constructor, but I'm worried we'll have a race condition when bootstrapping a network. For example on the `localnet` if the collection nodes are started before the access node (which is how it is currently configured), this will cause the collection node startup to fail if the access node doesn't happen to be up by the time we get to this point.

Reply:
Ahh yep, Instead I've created a method `BaseClient.GetAccount` which fetches the account and also checks the `AccountKeyIndex` is within range.